### ISSUE TITLE
true/false data structure prevents changing values

### ISSUE THREAD

**Issue by zhengwy888** [ingydotnet/yaml-libyaml-pm#25]
When loading an yaml with true value, the value can't be modified as perl will complain 'changing a read-only value' . to reproduce 

``` perl
use YAML::XS qw(Load);
my $yaml = <<EOM;
hello: true
EOM
my $data = Load($yaml);
$data->{hello}= 1; #<-- this line will error out with 'Modification of a read-only value attempted ...'

```

I don't know if there is a reason why converting to a data structure that Perl can't modify.

**Comment by ratsbane** [comment]
I'm not able to reproduce this problem or the output you're seeing.  The syntax around your heredoc doesn't look right to me - I think it should be something like this instead:
my $yaml = <<EOM;
 hello: true
EOM
my $data = Load($yaml);

**Comment by zhengwy888** [comment]
@ratsbane sorry markdown ate my formatting. I have updated the comment with the correct format. 
This is the perl version I am using:

```
Summary of my perl5 (revision 5 version 14 subversion 1) configuration:

  Platform:
    osname=linux, osvers=2.6.29.4, archname=x86_64-linux
    uname='linux l-c4u5build64 2.6.29.4 #8 smp tue may 25 14:00:33 pdt 2010 x86_64 x86_64 x86_64 gnulinux '
    config_args='-des -Dprefix=/home/utils/perl-5.14/5.14.1-nothreads-64'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='3.4.6 20060404 (Red Hat 3.4.6-8)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64 /usr/local/lib64
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.3.4.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.3.4'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib'

```

And this is the YAML::XS version
```
$YAML::XS::VERSION = '0.37';
```

**Comment by oalders** [comment]
I can reproduce this under perl 5.18.2

**Comment by dolmen** [comment]
Here is a minimalist test case (verified on perl 5.20.1):
```
$ perl -MYAML::XS -E 'Load("a: true")->{a} = !!0'
Modification of a read-only value attempted at -e line 1
```
This should not die.

**Comment by rurban** [comment]
Looks like I have to implement the same trick as in JSON::XS, special overloaded symbols representing true and false. 
A copy via `newSVsv(&PL_sv_yes);` helps with the readonly problem, but leads to lossage in the roundtrips:
```
test/boolean.t .......... 1/5
#   Failed test 'Booleans YNY roundtrip'
#   at test/boolean.t line 22.
#          got: '---
# a: 1
# b: 1
# c: ''
# d: ''
# '
#     expected: '---
# a: true
# b: 1
# c: false
# d: ''
# '
```

**Comment by mithun** [comment]
The workaround is to `delete` the key and reset it
```perl
use YAML::XS qw(Load);
my $yaml = <<EOM;
hello: true
EOM
my $data = Load($yaml);
# This fails
# $data->{hello}= 1;
# This works
delete $data->{hello};
$data->{hello} = 1;
```

**Comment by zhengwy888** [comment]
@rurban actually I think a round trip loss is not so much of an issue this case, since perl doesn't natively support true value so I don't think people expect that. plus we can have a flag to turn on/off this round trip value.
@mithun to avoid manually going into keys to find true/false I used JSON to dump than load the object to get rid of the true/false read only values.

**Comment by rurban** [comment]
perl supports sv_yes and sv_no via the well-known constructs `!!1` and `!!0`.

```
perl -MDevel::Peek -e'print Dump !!1'
SV = PVNV(0x7fb312003610) at 0x10433ba00
  REFCNT = 2147483644
  FLAGS = (IOK,NOK,POK,READONLY,PROTECT,pIOK,pNOK,pPOK)
  IV = 1
  NV = 1
  PV = 0x104300702 "1"
  CUR = 1
  LEN = 0
```
Note the overlarge refcount and PROTECT denoting an immortal protected special value, i.e. `&PL_sv_yes`.

**Comment by perlpunk** [comment]
Starting with YAML::XS 0.66_001, you can use this:
```perl
local $YAML::XS::Boolean = "JSON::PP"; # or "boolean"
my $data = Load("booltrue: true");
$data->{boolfalse} = JSON::PP::false;
my $yaml = Dump($data);
# boolfalse: false
# booltrue: true
```

See more about this in the docs. This also does not have the issue of readonly values.
I will close this issue, when 0.66_001 was tested by cpantesters and 0.67 was released.
Unless someone objects.
I was thinking about a third option "perl" that uses simple `1/''` instead of the special variables, but I'm not sure if this is really needed.

**Comment by SineSwiper** [comment]
I'd like to re-open this because this seems like the `load_scalar` code in `perl_libyaml.c` is doing this wrong: `scalar = &PL_sv_yes;`

Consider the following (using Perl 5.30):

```
$ perl -MDevel::Peek -e'Dump !!1'
SV = PVNV(0x558433cb0f80) at 0x558433caf3d8
  REFCNT = 2147483644
  FLAGS = (IOK,NOK,POK,READONLY,PROTECT,pIOK,pNOK,pPOK)
  IV = 1
  NV = 1
  PV = 0x558431d9b864 "1"
  CUR = 1
  LEN = 0
$ perl -MDevel::Peek -e'my $t = !!1; Dump $t'
SV = PVNV(0x55e954ed6040) at 0x55e954f04938
  REFCNT = 1
  FLAGS = (IOK,NOK,POK,pIOK,pNOK,pPOK)
  IV = 1
  NV = 1
  PV = 0x55e954f27130 "1"\0
  CUR = 1
  LEN = 10
```

One is `READONLY,PROTECT`, and the other is not.  I don't think the scalar can be set directly as a reference to the constant itself, like the code in `perl_libyaml.c`.  It needs a `newSV`, and that constant stashed into the new SV.

**Comment by perlpunk** [comment]
The current solution was chosen to be able to roundtrip booleans.
I've now created #118 which makes them not readonly for perl >= 5.36, where we got core boolean support.

**Comment by perlpunk** [comment]
#118 is now released as v0.902.0
